pkg/protocol/engine/attestation/slotattestation/manager.go

```diff
@@ -0,0 +1,178 @@
+	lastCumulativeWeight        uint64
+	attestationCommitmentOffset iotago.SlotIndex
+	commitmentMutex             sync.RWMutex
+
+	module.Module
+}
+
+func NewProvider() module.Provider[*engine.Engine, attestation.Attestation] {
+	return module.Provide(func(e *engine.Engine) attestation.Attestation {
+		m := NewManager(4, e.Storage.Prunable.Attestations, e.SybilProtection.Accounts)
+
+		e.HookInitialized(func() {
+			cw := e.Storage.Settings().LatestCommitment().Commitment().CumulativeWeight
+			index := e.Storage.Settings().LatestCommitment().ID().Index()
+			m.Initialize(index, cw)
+		})
+
+		return m
+	})
+}
+
+func NewManager(attestationCommitmentOffset iotago.SlotIndex, bucketedStorage func(index iotago.SlotIndex) kvstore.KVStore, weightsProviderFunc func() *account.Accounts[iotago.AccountID, *iotago.AccountID]) *Manager {
+	m := &Manager{
+		attestationCommitmentOffset: attestationCommitmentOffset,
+		weightsProviderFunc:         weightsProviderFunc,
+		bucketedStorage:             bucketedStorage,
+		pendingAttestations:         memstorage.NewIndexedStorage[iotago.SlotIndex, iotago.AccountID, *iotago.Attestation](),
+	}
+	m.tracker = NewTracker(m.computeAttestationCommitmentOffset)
+	m.TriggerConstructed()
+
+	return m
+}
+
+func (m *Manager) Initialize(lastCommitmentIndex iotago.SlotIndex, lastCumulativeWeight uint64) {
+	m.commitmentMutex.Lock()
+	m.lastCommittedSlot = lastCommitmentIndex
+	m.lastCumulativeWeight = lastCumulativeWeight
+	m.commitmentMutex.Unlock()
+
+	m.TriggerInitialized()
+}
+
+func (m *Manager) Shutdown() {
+	m.TriggerStopped()
+}
+
+func (m *Manager) AddAttestationFromBlock(block *blocks.Block) {
+	m.commitmentMutex.RLock()
+	defer m.commitmentMutex.RUnlock()
+
+	// We only care about attestations that are newer than the last committed slot.
+	if block.ID().Index() <= m.lastCommittedSlot {
+		return
+	}
+
+	// Attestations that are older than m.lastCommittedSlot - m.attestationCommitmentOffset don't have any effect, so we ignore them.
+	if block.SlotCommitmentID().Index() <= m.computeAttestationCommitmentOffset() {
+		return
+	}
+
+	newAttestation := iotago.NewAttestation(block.Block())
+
+	storage := m.pendingAttestations.Get(block.ID().Index(), true)
+	storage.Compute(block.Block().IssuerID, func(currentValue *iotago.Attestation, exists bool) *iotago.Attestation {
+		if !exists {
+			return newAttestation
+		}
+
+		// Replace the stored attestation only if the new one is greater.
+		if newAttestation.Compare(currentValue) == 1 {
+			return newAttestation
+		}
+
+		return currentValue
+	})
+}
+
+func (m *Manager) Commit(index iotago.SlotIndex) (newCW uint64, attestationsMap *ads.Map[iotago.AccountID, iotago.Attestation, *iotago.AccountID, *iotago.Attestation], err error) {
+	m.commitmentMutex.Lock()
+	defer m.commitmentMutex.Unlock()
+
+	// Remove all pending attestations of slot and apply to tracker.
+	pendingAttestations := m.pendingAttestations.Evict(index)
+	if pendingAttestations != nil {
```

Comment:
Yes if there's no attestation of a validator in the current slot. 

I don't think it changes anything. It obviously depends on the combination of parameters but an empty slot shouldn't have any bad consequences. If there's many slots without attestations it might result in no attestations for a slot that we're committing. But that is fine too, as the CW then simply doesn't grow and the tree will be empty. 

In fact, for the first slots that is even the case, as nodes first need to commit to genesis and only after `attestationCommitmentOffset` is reached we get attestations for the slot we're committing.